rmf_gazebo_plugins/src/readonly.cpp

```diff
@@ -0,0 +1,466 @@
+
+}
+
+void ReadonlyPlugin::map_cb(const BuildingMap::SharedPtr msg)
+{
+  if (msg->levels.empty())
+  {
+    RCLCPP_ERROR(logger(), "Received empty building map");
+    return;
+  }
+
+  _found_map = true;
+  _map = *msg;
+  RCLCPP_INFO(logger(), "Received building map with %d levels", msg->levels.size());
+  _found_level = false;
+  _found_graph = false;
+  for (const auto& level: msg->levels)
+  {
+    RCLCPP_INFO(logger(), "Level name: [%s]", level.name.c_str());
+    if (level.name.c_str() == _level_name)
+    {
+      _level = level;
+      _found_level = true;
+      RCLCPP_INFO(logger(), "Found level [%s] with %d nav_graphs",
+          level.name.c_str(), level.nav_graphs.size());
+
+      if (_nav_graph_index < level.nav_graphs.size())
+      {
+        _found_graph = true;
+        RCLCPP_INFO(logger(), "Graph index [%d] containts [%d] waypoints",
+            _nav_graph_index,
+            level.nav_graphs[_nav_graph_index].vertices.size());
+        initialize_graph();
+      }
+      else
+      {
+        RCLCPP_ERROR(logger(),
+            "Specified nav_graph index [%d] does not exist in level [%s]",
+            _nav_graph_index, _level_name.c_str());
+      }
+      break;
+    }
+  }
+
+  if (!_found_level)
+    RCLCPP_ERROR(logger(),
+        "Did not find level [%s] in building map. Path will not be published.",
+        _level_name.c_str());
+}
+
+void ReadonlyPlugin::initialize_graph()
+{
+  if (!_found_graph)
+    return;
+  
+  std::lock_guard<std::mutex> lock(_mutex);
+
+  _initialized_graph = false;
+
+  const auto start_time = std::chrono::steady_clock::now();
+  _graph = _level.nav_graphs[_nav_graph_index];
+  RCLCPP_INFO(logger(), "Nav graph contains [%d] lanes" , _graph.edges.size());
+  for (const auto& edge : _graph.edges)
+  {
+    // Inserting entry for v1_idx
+    auto entry = _neighbor_map.find(edge.v1_idx);
+    if (entry == _neighbor_map.end())
+    {
+      // We assume the vertex is a neighbor of itself to account for when the
+      // robot turns around and heads back to the start index
+      std::unordered_set<std::size_t> neighbors({edge.v1_idx, edge.v2_idx});
+      _neighbor_map.insert(std::make_pair(edge.v1_idx, neighbors));
+    }
+    // Updating existing entry
+    else
+    {
+      entry->second.insert(edge.v2_idx);
+    }
+
+    // If bidrectional, add entry for v2_idx
+    if (edge.edge_type == edge.EDGE_TYPE_BIDIRECTIONAL)
+    {
+      entry = _neighbor_map.find(edge.v2_idx);
+      if (entry == _neighbor_map.end())
+      {
+        std::unordered_set<std::size_t> neighbors({edge.v1_idx, edge.v2_idx});
+        _neighbor_map.insert(std::make_pair(edge.v2_idx, neighbors));
+      }
+      // Updating existing entry
+      else
+      {
+        entry->second.insert(edge.v1_idx);
+      }
+    }
+
+  }
+
+  // Initialize _wp_names
+  for (std::size_t i =0; i < _graph.vertices.size(); ++i)
+    _wp_names.insert(std::make_pair(i, _graph.vertices[i].name));
```

Comment:
`_wp_names[i]` seems to serve the same purpose as `_graph.vertices[i]`, we can probably do away with this `unordered_map`

Reply:
You're right. In a previous version I only stored the waypoints which had non-empty names into this unordered_map. But now its redundant.